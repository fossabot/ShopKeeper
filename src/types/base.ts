// Copyright (C) 2018 Sleep E-Z USA, Inc. / Evan Darwin
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {Required} from '../lib/typing';
import {ProductImageItem} from './shopify/image';
import {Metafield} from './shopify/metafield';
import {Store} from './store';

/**
 * These types define special strings used for reading Shopify
 * responses, as well as various other uses throughout the
 * system. Each [[ShopifyType]] class must provide an `itemType`
 * containing one of these enum values.
 */
export enum ShopifyTypes {
  Page = 'pages',
  Option = 'product_option',
  Variant = 'variants',
  ProductImage = 'product_image',

  Product = 'products',
  Country = 'countries',
  Metafield = 'metafield',
  Province = 'province',
  Redirect = 'redirect',
  Shop = 'shop',
}

/**
 * The ShopifyType is one of the foundational types for ShopKeeper.
 * All entities that can be managed with the ShopKeeper framework
 * must extend this class to maintain simplicity and interoperability.
 *
 * @typeparam R   This generic describes the shape of the incoming data
 *                as returned from Shopify's Admin API.
 *
 * @typeparam U   This generic describes the shape of the incoming data
 *                as provided locally by the user (it may not exist remotely)
 */
export abstract class ShopifyItem<R extends RawShopifyData,
                                            U extends UserData> {
  abstract itemType: ShopifyTypes;
  includesRemote: boolean;
  store: Store;
  data: Indexable&Partial<{handle: Handle}>&Required<U>;
  protected rawData: R;
  protected userData: U|null;

  constructor(store: Store, data: R, remote = false) {
    this.store = store;
    this.includesRemote = remote;
    this.rawData = data;
  }

  id(): number {
    return this.rawData.id || this.data.id || -1;
  }

  isResolved(): boolean {
    return (this.id() > -1);
  }

  addUserData(data: U) {
    this.userData = data;
  }

  toShopifyJSON(): R {
    return this.mergeToRaw();
  }

  abstract mergeToRaw(): R;
}

export abstract class LocalShopifyItem<D extends UserData> {
  abstract itemType: ShopifyTypes;
  abstract data: D;

  constructor(data: D) {
    this.data = data;
  }
}

/**
 * This is data given by the user to re-create the user's
 * intended Shopify item. This means that this type will
 * likely have optional fields as the user should be allowed
 * to define properties as they please (except required fields).
 *
 * **NOTE**: If the `id` field is provided in the user's config,
 * it will simply be ignored and replaced by the remote value.
 * Concurrently, the user should *never* define the `id` value
 * as it will differ between environments.
 */
export type UserData = {};

/**
 * This type is used to represent the raw JSON response as received
 * from Shopify. Since it is returned from Shopify, this **`id`**
 * value will be the preferred reference method.
 */
export type RawShopifyData = Indexable&{};
export type ShopifyData = Indexable&UserData;

export interface ShopifyType<D, I> { new(store: Store, data: D): I; }

/**
 * We specify a type to allow uniqueness among
 * types carrying a `handle` property.
 */
export type Handle = string;

/**
 * All Shopify types available in the scope of
 * ShopKeeper are referenced based on their ID
 * available from Shopify.
 */
export type Indexable = {
  id: number,
};

/**
 * The following Shopify types carry the 'handle'
 * element: Article, Blog, CustomCollection,
 *          SmartCollection, Page, Product, Blog
 */
export type Handleable<H extends Handle> = {
  handle: H,
};

export type Purchasable = {
  price?: number,
  comparePrice?: number
};

export type Purchasable$Raw = {
  price: number|null,
  compare_at_price: number|null
};

/**
 * The Shopify type provides a `title` field for
 * the name of the element. If available, the field
 * is always required.
 */
export type Titled = {
  title: string
};

export type Metafields = {
  metafields: Array<Metafield<string>>
};
export type Metafields$Raw = Metafields;

/**
 * The type is positional and thus can be given a
 * specified position during submission. No actions
 * are required in the user's projects to position
 * these besides re-ordering their array elements.
 */
export type Positional = Partial<Positional$Raw>;
export type Positional$Raw = {
  position: number
};

/**
 * The element carries a timestamp of when the
 * element was created.
 */
export type Date$CreatedAt = {
  createdAt: Date
};
export type Date$CreatedAt$Raw = {
  created_at: string
};

/**
 * The element carries a timestamp of when the
 * element was last updated. This value can
 * sometimes be modified after the fact.
 */
export type Date$UpdatedAt = {
  updatedAt: Date
};
export type Date$UpdatedAt$Raw = {
  updated_at: string
};

/**
 * The element carries a timestamp of when the
 * element was last published. This value can be
 * edited after the fact.
 */
export type Date$PublishedAt = {
  publishedAt?: Date|null
};
export type Date$PublishedAt$Raw = {
  published_at: string|null
};

/**
 * The element accepts simple HTML content in
 * the form of a string.
 */
export type Content$HTML = {
  html?: string|'',
  template?: string|null,
};

export type Content$HTML$Raw = {
  body_html: string
};

/**
 * The element accepts binary or other formatted data
 * and can be submitted to Shopify's servers
 */
export type Content$Binary = {
  path?: string,
  filename?: string
};
export type Content$Binary$Raw = {
  attachment?: string,
  filename?: string
};

/**
 * If the element allows for HTML we can specify a
 * template to render and replace the `body_html`
 * content. If the `template` parameter is specified
 * then the `body_html` will be overridden.
 */
export type Content$Template = {
  template?: string
};
export type Content$Template$Raw = {
  body_html: string
};

export type Image$Single = {
  image?: ProductImageItem|null
};
export type Image$Single$Raw = {
  image_id?: number
};

export type Image$Multiple = {
  images?: ProductImageItem[]
};
export type Image$Multiple$Raw = {
  images?: number[]
};