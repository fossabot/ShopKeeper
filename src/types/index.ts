export * from './base';
export * from './cache';
export * from './config';
export * from './inventory';
export * from './routes';
export * from './store';
export * from './shopify/index';
