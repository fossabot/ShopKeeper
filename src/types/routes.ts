// Copyright (C) 2018 Sleep E-Z USA, Inc. / Evan Darwin
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {ShopifyTypes} from './base';

export enum Method {
  List = 'list',
  Count = 'count',
  Fetch = 'fetch',
  Create = 'create',
  Update = 'update',
  Delete = 'delete',
}

export type ItemRouteMethods = {
  list: string|null,
  count: string|null,
  fetch: string|null,
  create: string|null,
  update: string|null,
  delete: string|null,
  [method: string]: string|null,
};

export const STANDARD_ITEM_ROUTE_METHODS: ItemRouteMethods = {
  list: '{prefix}.json',
  create: '{prefix}.json',
  count: '{prefix}/count.json',
  fetch: '{prefix}/{id}.json',
  update: '{prefix}/{id}.json',
  delete: '{prefix}/{id}.json',
};

export const STANDARD_ITEM_ROUTES: ItemRoutes = {
  paginated: false,
  publicPath: null,
  methods: STANDARD_ITEM_ROUTE_METHODS,
  prefix: '',
};

export type ItemRoutes = {
  /**
   * The prefix is the type as documented by Shopify's Admin
   * API. For example the PriceRule type's prefix would become
   * 'price_rule'. This is used for automatic URL resolution.
   */
  prefix: string,

  // This is a string with supported variables that will
  // automatically be replaced by ShopKeeper's engine.
  publicPath: string|null,

  // Describes which methods are available and provides
  // additional context to support Shopify's complex routing
  methods: ItemRouteMethods,

  // Describes if retrieval of all objects requires
  // pagination via the REST API.
  paginated: boolean,
};

export const GLOBAL_SHOPKEEPER_ROUTES: {[name: string]: ItemRoutes} = {
  [ShopifyTypes.Country]: {
    ...STANDARD_ITEM_ROUTES,
    prefix: 'countries',
  },

  [ShopifyTypes.Province]: {
    ...STANDARD_ITEM_ROUTES,
    prefix: 'provinces',
    methods: {
      update: 'countries/{country_id}/provinces/{id}.json',
      fetch: 'countries/{country_id}/provinces/{id}.json',
      list: 'countries/{country_id}/provinces.json',
      count: 'countries/{country_id}/provinces/count.json',
      create: null,
      delete: null,
    }
  },

  [ShopifyTypes.Product]: {
    ...STANDARD_ITEM_ROUTES,
    publicPath: '{prefix}/{handle}',
    prefix: 'products',
    paginated: true,
  },

  [ShopifyTypes.Variant]: {
    ...STANDARD_ITEM_ROUTES,
    publicPath: 'products/{handle}?variant={id}',
    prefix: 'variants',
    paginated: false,
    methods: {
      ...STANDARD_ITEM_ROUTE_METHODS,
      list: 'products/{pid}/{prefix}.json',
      count: 'products/{pid}/{prefix}/count.json',
      delete: 'products/{pid}/{prefix}/delete.json',
    },
  },

  [ShopifyTypes.Redirect]: {
    ...STANDARD_ITEM_ROUTES,
    prefix: 'redirects',
    publicPath: '{origin}',
    paginated: true,
  },

  [ShopifyTypes.Shop]: {
    ...STANDARD_ITEM_ROUTES,
    prefix: 'shop',
    methods: {
      ...STANDARD_ITEM_ROUTE_METHODS,
      fetch: '{prefix}.json',
      count: null,
      create: null,
      update: null,
      delete: null,
      list: null,
    },
  },

  [ShopifyTypes.Page]: {
    ...STANDARD_ITEM_ROUTES,
    prefix: 'pages',
    methods: STANDARD_ITEM_ROUTE_METHODS,
  },
};