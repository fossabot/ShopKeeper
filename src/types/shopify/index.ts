export * from './country';
export * from './image';
export * from './metafield';
export * from './option';
export * from './page';
export * from './product';
export * from './province';
export * from './redirect';
export * from './shop';
export * from './variant';