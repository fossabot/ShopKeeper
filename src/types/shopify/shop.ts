// Copyright (C) 2018 Sleep E-Z USA, Inc. / Evan Darwin
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {Indexable, ShopifyItem, ShopifyTypes} from '../base';
import {Store} from '../store';

export class ShopItem extends ShopifyItem<ShopData$Raw, ShopData$Raw> {
  itemType = ShopifyTypes.Shop;

  constructor(store: Store, data: ShopData$Raw) {
    super(store, data);
    this.data = data;

    store.registerToCache(this);
  }

  mergeToRaw(): ShopData$Raw {
    return this.rawData;
  }
}

export type ShopData$Raw = Indexable&{
  name: string,
  email: string,
  domain: string,
  province: string,
  country: string,
  address1: string,
  address2: string,
  zip: string,
  city: string,
  source: string | null,
  phone: string,
  customer_email: string,
  latitude: number,
  longitude: number,
  primary_location_id: number | null,
  primary_locale: string,
  country_code: string,
  country_name: string,
  currency: string,
  timezone: string,
  iana_timezone: string,
  shop_owner: string,
  money_format: string,
  money_with_currency_format: string,
  weight_unit: string,
  province_code: string,
  taxes_included: null | boolean,
  tax_shipping: null | boolean,
  county_taxes: boolean,
  plan_display_name: string,
  plan_name: string,
  has_discounts: boolean,
  has_gift_cards: boolean,
  myshopify_domain: string,
  google_apps_domain: null | string,
  google_apps_login_enabled: null | boolean,
  money_in_emails_format: string,
  money_with_currency_in_emails_format: string,
  eligible_for_payments: boolean,
  requires_extra_payments_agreement: boolean,
  password_enabled: boolean,
  has_storefront: boolean,
  eligible_for_card_reader_giveaway: boolean | null,
  finances: boolean,
  setup_required: boolean,
  force_ssl: boolean,
};