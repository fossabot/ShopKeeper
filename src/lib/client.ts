// Copyright (C) 2018 Sleep E-Z USA, Inc. / Evan Darwin
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import * as CoreTypes from '../types/base';
import {Handle, LocalShopifyItem, ShopifyTypes} from '../types/base';
import {OptionData$Raw, OptionData$User} from '../types/shopify/option';
import {ProductData$LocalOnly, ProductData$User} from '../types/shopify/product';
import {VariantData$User} from '../types/shopify/variant';


export type Lib$ProductOptions = Array<{name: string, values: string[]}>;

export function generateProductOptions(options: Lib$ProductOptions) {
  const finalOptions: OptionData$Raw[] = [];
  const modifiedOptions = {...options};

  if (Object.keys(options).length > 3) {
    throw new Error(
        'Three option dimensions is the maximum allowed ' +
        'on the Shopify platform');
  }
}

export class Product<H extends Handle> extends
    LocalShopifyItem<ProductData$LocalOnly<H>> {
  itemType = ShopifyTypes.Product;
  data: ProductData$LocalOnly<H>;
}

export class Option<H extends Handle> extends
    LocalShopifyItem<OptionData$User> {
  itemType = ShopifyTypes.Option;
  data: OptionData$User;
}

export class Variant<H extends Handle> extends
    LocalShopifyItem<VariantData$User<H>> {
  itemType = ShopifyTypes.Variant;
  data: VariantData$User<H>;
}


export type Helpers$MakeOptions$Result = {
  unresolvedOptions: Array<Option<Handle>>; options: never[];
};
export type Helpers$MakeOptions$Option = {
  name: string; values: string[];
};
export function makeOptions(...options: Helpers$MakeOptions$Option[]):
    Helpers$MakeOptions$Result {
  if (options.length > 3) {
    throw new Error(`makeOptions Error: Maximum allowed is three (3)`);
  }

  if (options.length === 0) {
    // TODO: Possibly prefer to create the Shopify default instead?
    throw new Error(`makeOptions Error: You must provide at least one option`);
  }

  return {
    unresolvedOptions: options.map((option) => new Option(option)),
    options: [],
  };
}

export type Helpers$MakeVariants$Callback = (options: string[]) =>
    Variant<Handle>;
export type Helpers$MakeVariants$Result = {
  unresolvedVariants: Array<Variant<Handle>>; variants: never[];
};

export function makeVariants(
    options: Array<Option<Handle>>,
    cb: Helpers$MakeVariants$Callback): Helpers$MakeVariants$Result {
  // TODO: Remove me, makes missing params shut up
  if (options) {
    (false) ? cb([]) : cb([]);
  }

  return {
    unresolvedVariants: [],
    variants: [],
  };
}