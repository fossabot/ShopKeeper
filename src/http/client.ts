// Copyright (C) 2018 Sleep E-Z USA, Inc. / Evan Darwin
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {RequestCache, RequestMode} from 'node-fetch';

import {getOrFail} from '../lib/typing';
import {shopify_printf} from '../routing';
import {colorType} from '../theme';
import * as CoreTypes from '../types/base';
import {GLOBAL_SHOPKEEPER_ROUTES, ItemRoutes, Method} from '../types/routes';
import {Country$Raw, CountryItem} from '../types/shopify/country';
import {PageData$Raw, PageItem} from '../types/shopify/page';
import {ProductData$Raw, ProductItem} from '../types/shopify/product';
import {ShopData$Raw, ShopItem} from '../types/shopify/shop';
import {Store} from '../types/store';

export type GenericApiResponse = {
  [type: string]: {}|Array<{}>
};

export type RequestParams = {};
export type FetchRequestMethod = 'GET'|'POST'|'PUT'|'DELETE';
export type FetchOptions = {
  method: FetchRequestMethod,
  mode?: RequestMode,
  cache?: RequestCache
  headers?: {[handle: string]: string},
};

  export const DEFAULT_FETCH_OPTS: FetchOptions = {
    method: 'GET',
    mode: 'cors',
    cache: 'default'
  };


  export type GenericShopifyType =
      CoreTypes.ShopifyItem<CoreTypes.RawShopifyData, CoreTypes.UserData>;
  export class ShopifyClient {
    /**
     * Note: This creates a new Object as to not use a reference,
     *       we modify this value during testing.
     */
    routes: {[name: string]: ItemRoutes} =
        Object.assign({}, GLOBAL_SHOPKEEPER_ROUTES);
    options: FetchOptions;
    store: Store;

    constructor(store: Store, options?: Partial<FetchOptions>) {
      this.store = store;

      const userOptions = options;
      const headers = Object.assign({}, DEFAULT_FETCH_OPTS.headers,
        (userOptions ? userOptions.headers : {}),
        {'Authorization': `Basic ${this._basic_auth_hash()}`});

      this.options = {
        ...DEFAULT_FETCH_OPTS,
        ...userOptions,
        headers,
      };
    }

    _request<R extends GenericShopifyType>(
        method: Method, type: CoreTypes.ShopifyTypes): Promise<R[]> {
      const methodVerb = {
        [Method.List]: 'GET',
        [Method.Fetch]: 'GET',
        [Method.Count]: 'GET',
        [Method.Create]: 'POST',
        [Method.Update]: 'PUT',
        [Method.Delete]: 'DELETE',
      }[method];

      const routes = this.routes[type];
      if (!routes) {
        throw new Error(
            `Refusing to fulfill ${methodVerb} request: ` +
            `No routes available for type '${colorType(type)}'`);
      }

      const requestOpts = {
        ...this.options,
        method: methodVerb,
      };

      const methodUrl = routes.methods[method];
      if (!methodUrl) {
        throw new Error(
            `Refusing to fulfill ${methodVerb} request: ` +
            `Method not allowed for type ${type}`);
      }

      const url = [
        `https://${this.store.data.url}/admin`,
        shopify_printf(methodUrl, {prefix: routes.prefix}),
      ].join('/');

      return fetch(url, requestOpts as RequestInit)
          .then((res) => {
            if (!res.ok) {
              return Promise.reject(new Error(
                  `Request Failed: ${res.status} (${res.statusText})`));
            }

            return res.json();
          })
          .then(
              (data: GenericApiResponse) =>
                  (this.process_results(type, data) as R[]));
    }

    create<T extends GenericShopifyType>(itemType: CoreTypes.ShopifyTypes):
        Promise<T[]|null> {
      return this._request<T>(Method.Create, itemType);
    }

    update<T extends GenericShopifyType>(itemType: CoreTypes.ShopifyTypes):
        Promise<T[]|null> {
      return this._request<T>(Method.Update, itemType);
    }

    delete<T extends GenericShopifyType>(itemType: CoreTypes.ShopifyTypes):
        Promise<T[]|null> {
      return this._request(Method.Delete, itemType);
    }

    fetch<T extends GenericShopifyType>(itemType: CoreTypes.ShopifyTypes):
        Promise<T[]|null> {
      return this._request(Method.Fetch, itemType);
    }

    count<T extends GenericShopifyType>(itemType: CoreTypes.ShopifyTypes):
        Promise<T[]|null> {
      return this._request(Method.Count, itemType);
    }

    list<T extends GenericShopifyType>(itemType: CoreTypes.ShopifyTypes):
        Promise<T[]|null> {
      return this._request(Method.List, itemType);
    }

    process_result(
        type: CoreTypes.ShopifyTypes,
        data: CoreTypes.RawShopifyData): GenericShopifyType {
      if (type === CoreTypes.ShopifyTypes.Country) {
        return new CountryItem(this.store, data as Country$Raw<string>);
      }

      if (type === CoreTypes.ShopifyTypes.Product) {
        return new ProductItem(
            this.store, data as ProductData$Raw<CoreTypes.Handle>);
      }

      if (type === CoreTypes.ShopifyTypes.Shop) {
        return new ShopItem(this.store, data as ShopData$Raw);
      }

      if (type === CoreTypes.ShopifyTypes.Page) {
        return new PageItem(this.store, data as PageData$Raw<CoreTypes.Handle>);
      }

      throw new Error(
          `Failed to process: Unknown or unprocessable response given`);
    }
    process_results<R extends GenericShopifyType>(
        type: CoreTypes.ShopifyTypes, data: GenericApiResponse): R[] {
      if (!data) {
        throw new Error(`Failed to process Shopify results for type ${type}`);
      }

      const resultItems: CoreTypes.RawShopifyData|CoreTypes.RawShopifyData[] =
          getOrFail(data[type]);

      return (Array.isArray(resultItems) ? resultItems : [resultItems])
          .map((item) => this.process_result(type, item) as R);
    }

    _basic_auth_hash(): string {
      const secrets = this.store.data.secrets;
      const hashStr = `${secrets._key}:${secrets._password}`;

      return (new Buffer(hashStr)).toString('base64');
    }
  }