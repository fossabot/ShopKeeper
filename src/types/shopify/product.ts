// Copyright (C) 2018 Sleep E-Z USA, Inc. / Evan Darwin
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {Option, Variant} from '../../lib/client';
import {dateOrUndefined, dateToShopify, getOrFail, getOrNull, getOrUndefined} from '../../lib/typing';
import * as CoreTypes from '../base';
import {Store} from '../store';

import {ProductImageItem} from './image';
import {OptionData$Raw, OptionItem} from './option';
import {Variant$Raw, VariantItem} from './variant';


export type ProductData$User<H extends CoreTypes.Handle> =
    CoreTypes.Handleable<H>&CoreTypes.Content$HTML&{
  title: string;
  price?: number|null;
  comparePrice?: number|null;

  createdAt?: Date, updatedAt?: Date, publishedAt?: Date,

      image?: ProductImageItem|null, images?: ProductImageItem[]|never[],

      vendor?: string;
  productType?: string;
  templateSuffix?: string;
  publishedScope?: string;
  tags?: string[];
  variants: Array<VariantItem<H>>;
  options: Array<OptionItem<H>>;
};

export type ProductData$LocalOnly<H extends CoreTypes.Handle> =
    ProductData$User<H>&{
  unresolvedVariants: Array<Variant<H>>;
  unresolvedOptions: Array<Option<H>>;
};

export type ProductData$Raw<H extends CoreTypes.Handle> =
    CoreTypes.Indexable&CoreTypes.Titled&CoreTypes.Purchasable$Raw&
    CoreTypes.Image$Single$Raw&CoreTypes.Image$Multiple$Raw&
    CoreTypes.Date$CreatedAt$Raw&CoreTypes.Date$UpdatedAt$Raw&
    CoreTypes.Date$PublishedAt$Raw&CoreTypes.Content$HTML$Raw&{
  handle: H;

  variants: Array<Variant$Raw<H>>;
  options: OptionData$Raw[];

  tags: string|null;
  vendor: string|null;
  product_type: string|null;
  template_suffix: string|null;
  published_scope: string;
};

export class ProductItem<H extends CoreTypes.Handle> extends
    CoreTypes.ShopifyItem<ProductData$Raw<H>, ProductData$User<H>> {
  itemType = CoreTypes.ShopifyTypes.Product;

  constructor(store: Store, data: ProductData$Raw<H>, remote = false) {
    super(store, data, remote);

    const {
      id,
      title,
      handle,
      product_type,
      vendor,
      options: rawOptions,
      variants: rawVariants,
      body_html,
      image_id,
      images,
      price,
      compare_at_price: comparePrice,
      created_at,
      updated_at,
      published_at,
      published_scope: publishedScope,
      tags,
      template_suffix: templateSuffix
    } = data;

    const options = rawOptions.map((option) => {
      return new OptionItem<H>(store, option);
    });

    const variants = rawVariants.map((variant) => {
      return new VariantItem<H>(store, variant, this);
    });

    this.data = {
      id,
      title,
      handle,
      productType: getOrUndefined(product_type),
      vendor: getOrUndefined(vendor),

      // TODO
      options,
      variants,

      template: null,
      html: body_html,

      // TODO: Proper image loading
      image: null,
      images: [],

      tags: (tags === undefined || !tags) ? undefined :
                                            tags.split(',').map(t => t.trim()),

      price: getOrUndefined<number>(price),
      comparePrice: getOrUndefined<number>(comparePrice),

      createdAt: dateOrUndefined(created_at),
      updatedAt: dateOrUndefined(updated_at),
      publishedAt: dateOrUndefined(published_at),

      publishedScope,
      templateSuffix: getOrUndefined(templateSuffix),
    };

    store.registerToCache(this);
  }

  mergeToRaw(): ProductData$Raw<H> {
    const {
      id,
      title,
      handle,
      productType,
      vendor,
      price,
      comparePrice,
      template,
      html,
      createdAt,
      updatedAt,
      publishedAt,
      publishedScope,
      variants,
      options,
      tags,
      templateSuffix,
    } = this.data;

    const renderedContent = (!!template) ? template : html || '';

    const rawResult: ProductData$Raw<H> = {
      id,
      title,
      handle,

      tags: (tags || ['']).join(','),
      body_html: renderedContent,

      vendor: getOrNull(vendor),
      price: getOrNull(price),
      compare_at_price: getOrNull(comparePrice),
      product_type: getOrNull(productType),

      created_at: dateToShopify(getOrFail(createdAt)),
      updated_at: dateToShopify(getOrFail(updatedAt)),
      published_at: (!publishedAt) ? null : dateToShopify(publishedAt),

      variants: [],
      options: [],

      // TODO: Use global defaults here
      published_scope: getOrNull(publishedScope) || 'global',
      template_suffix: getOrNull(templateSuffix),
    };

    return rawResult;
  }
}
