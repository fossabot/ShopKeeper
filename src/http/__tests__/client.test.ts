import {Array} from 'core-js/library/web/timers';

import {testCreateProduct, testCreateProductRaw, testCreateStore} from '../../testing/data';
import {Handle, ShopifyType, ShopifyTypes} from '../../types/base';
import {Method} from '../../types/routes';
import {ProductItem} from '../../types/shopify/product';
import {ShopItem} from '../../types/shopify/shop';
import {Store} from '../../types/store';
import {DEFAULT_FETCH_OPTS, ShopifyClient} from '../client';

const fetchMock = require('fetch-mock');
const expectedFetchOpts = {
  method: 'GET',
  mode: 'cors',
  cache: 'default',
  headers: {Authorization: 'Basic Og=='},
};

describe('Shopify HTTP Client', () => {
  it('DEFAULT_FETCH_OPTS has not changed', () => {
    expect(DEFAULT_FETCH_OPTS).toMatchSnapshot();
  });

  describe('Basic Authentication', () => {
    it('generates properly', () => {
      const store = testCreateStore({
        secrets: {
          _key: 'something-with',
          _password: '182948-decent-entropy',
        }
      });

      const client = new ShopifyClient(store);
      expect(client._basic_auth_hash())
          .toEqual('c29tZXRoaW5nLXdpdGg6MTgyOTQ4LWRlY2VudC1lbnRyb3B5');
    });
  });

  describe('ShopifyClient', () => {
    const store = testCreateStore();

    it('should instantiate', () => {
      const client = new ShopifyClient(store);

      expect(client).toBeTruthy();
      expect(client).toBeInstanceOf(ShopifyClient);
    });

    it('should merge user options properly', () => {
      const client = new ShopifyClient(store, {
        cache: 'no-cache',
        headers: {
          Important: 'Auth-Header',
          Authorization: 'INVALID!'  // it will *always* overwrite this header
        },
        mode: 'no-cors',
      });

      expect(client.options).toEqual({
        cache: 'no-cache',
        method: 'GET',
        headers: {
          Authorization: 'Basic Og==',
          Important: 'Auth-Header',
        },
        mode: 'no-cors',
      });
    });

    it('refuses to fulfill when given an unknown type', () => {
      const client = new ShopifyClient(store);

      expect(() => {
        client._request(Method.Fetch, 'ADASDASD' as ShopifyTypes);
      }).toThrowErrorMatchingSnapshot();
    });

    it('refuses to fulfill when routes cannot be determined', () => {
      const client = new ShopifyClient(store);
      delete client.routes[ShopifyTypes.Product];

      expect(() => {
        client.fetch(ShopifyTypes.Product);
      }).toThrowErrorMatchingSnapshot();
    });

    it('_request() works properly', (done) => {
      const client = new ShopifyClient(testCreateStore());
      fetchMock.get('https://localhost/admin/shop.json', {shop: {id: 1}});

      client._request<ShopItem>(Method.Fetch, ShopifyTypes.Shop)
          .then((result: ShopItem[]|null) => {
            const lastCall = fetchMock.lastCall();

            expect(lastCall[0]).toEqual('https://localhost/admin/shop.json');
            expect(lastCall[1]).toEqual({
              method: 'GET',
              mode: 'cors',
              cache: 'default',
              headers: {Authorization: 'Basic Og=='},
            });

            expect(result).toHaveLength(1);
            expect((result as ShopItem[])[0]).toBeInstanceOf(ShopItem);

            done();
          });
    });

    it('list() works properly', (done) => {
      const store = testCreateStore();
      const client = new ShopifyClient(store);

      fetchMock.reset();
      fetchMock.get('https://localhost/admin/products.json', {
        products: [
          testCreateProductRaw({id: 1, handle: 'example1'}),
          testCreateProductRaw({id: 2, handle: 'example2'}),
        ]
      });

      client.list<ProductItem<Handle>>(ShopifyTypes.Product)
          .then((result: Array<ProductItem<Handle>>|null) => {
            const lastCall = fetchMock.lastCall();

            expect(result).toHaveLength(2);
            expect((result as Array<ProductItem<Handle>>)[0])
                .toBeInstanceOf(ProductItem);
            expect((result as Array<ProductItem<Handle>>)[1])
                .toBeInstanceOf(ProductItem);
            expect(lastCall[0])
                .toEqual('https://localhost/admin/products.json');
            expect(lastCall[1]).toEqual(expectedFetchOpts);

            done();
          });
    });
  });
});