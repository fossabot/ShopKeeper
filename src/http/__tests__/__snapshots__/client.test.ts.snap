// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopify HTTP Client DEFAULT_FETCH_OPTS has not changed 1`] = `
Object {
  "cache": "default",
  "method": "GET",
  "mode": "cors",
}
`;

exports[`Shopify HTTP Client ShopifyClient refuses to fulfill when given an unknown type 1`] = `"Refusing to fulfill GET request: No routes available for type 'ADASDASD'"`;

exports[`Shopify HTTP Client ShopifyClient refuses to fulfill when routes cannot be determined 1`] = `"Refusing to fulfill GET request: No routes available for type 'products'"`;
